thetis/rungekutta.py

```diff
@@ -0,0 +1,949 @@
+        """
+        if self._nontrivial:
+            if update_forcings is not None:
+                update_forcings(t + self.c[i_stage]*self.dt)
+            self.solver[i_stage].solve()
+
+    def get_final_solution(self, additive=False):
+        if not additive:
+            self.solution.assign(self.solution_old)
+        if self._nontrivial:
+            for j in range(self.n_stages):
+                self.solution += float(self.b[j])*self.tendency[j]
+        self.solution_old.assign(self.solution)
+
+    def solve_stage(self, i_stage, t, update_forcings=None):
+        self.update_solution(i_stage)
+        self.solve_tendency(i_stage, t, update_forcings)
+
+
+class ERKGenericShuOsher(TimeIntegrator):
+    """
+    Generic explicit Runge-Kutta time integrator.
+
+    Implements the Shu-Osher form.
+    """
+
+    def __init__(self, equation, solution, fields, dt, bnd_conditions=None, solver_parameters={}, terms_to_add='all'):
+        """Creates forms for the time integrator"""
+        super(ERKGenericShuOsher, self).__init__(equation, solution, fields, dt, solver_parameters)
+
+        self.tendency = Function(self.equation.function_space, name='tendency')
+        self.stage_sol = []
+        for i in range(self.n_stages):
+            s = Function(self.equation.function_space, name='sol'.format(i))
+            self.stage_sol.append(s)
+
+        # fully explicit evaluation
+        self.a_rk = self.equation.mass_term(self.equation.trial)
+        self.l_rk = self.dt_const*self.equation.residual(terms_to_add,
+                                                         self.solution, self.solution,
+                                                         self.fields, self.fields,
+                                                         bnd_conditions)
+        self._nontrivial = self.l_rk != 0
+        self.update_solver()
+
+    def update_solver(self):
+        if self._nontrivial:
+            prob = LinearVariationalProblem(self.a_rk, self.l_rk, self.tendency)
+            self.solver = LinearVariationalSolver(prob, options_prefix=self.name + '_k',
+                                                  solver_parameters=self.solver_parameters)
+
+    def initialize(self, solution):
+        pass
+
+    def solve_stage(self, i_stage, t, update_forcings=None):
+        if self._nontrivial:
+            if update_forcings is not None:
+                update_forcings(t + self.c[i_stage]*self.dt)
+
+            if i_stage == 0:
+                self.stage_sol[0].assign(self.solution)
+
+            # solve tendency
+            self.solver.solve()
+
+            # solve the next internal solution
+            # BUG starting with sol_sum = 0 results in different solution!
+            sol_sum = float(self.beta[i_stage + 1][i_stage])*self.tendency
+            for j in range(i_stage + 1):
+                sol_sum += float(self.alpha[i_stage + 1][j])*self.stage_sol[j]
+            self.solution.assign(sol_sum)
```

Comment:
I think I know what's going on (maybe).  This kind of formulation is intermingling symbolic computation and not.  UFL does some simplification of `a*b` if either of `a` or `b` are scalar values (or a symbolic zero).  In particular, if `a == 1`, then `1*b => b`.  In this case you assign the coefficient `self.tendency` to `sol_sum` and each intermediate `+=` now does the increment numerically into `self.tendency`, which is somewhat unsafe?

Moreover, you make more `par_loops` than necessary.

I would rewrite this as:

```python
import operator
sol_expr = reduce(operator.add, map(operator.mul, self.alpha[i_stage + 1], self.stage_sol), self.beta[i_stage+1][i_stage]*self.tendency)
self.solution.assign(sol_expr)
```

Or, if you don't like the functional style:

```python
sol_sum = float(self.beta[i_stage + 1][i_stage])*self.tendency
for j in range(i_stage + 1):
    sol_sum = sol_sum + float(self.alpha[i_stage + 1][j])*self.stage_sol[j]
self.solution.assign(sol_sum)
```
This is not equivalent to your code, because `__iadd__` is overridden differently from `__add__`.